katdal/test/test_lazy_indexer.py

```diff
@@ -73,19 +74,182 @@ def test_too_many_axes(self):
         self._test_index_error(np.s_[0, 0, 0, 0])
 
 
+def slice_to_range(s, l):
+    return range(*s.indices(l))
+
+
+def ix_(keep, shape):
+    """Extend numpy.ix_ to accept slices and single ints as well."""
+    # Inspired by Zarr's indexing.py (https://github.com/zarr-developers/zarr)
+    keep = [slice_to_range(k, s) if isinstance(k, slice)
+            else [k] if _is_integer(k)
+            else k
+            for k, s in zip(keep, shape)]
+    return np.ix_(*keep)
+
+
+def numpy_oindex(x, keep):
+    """Perform outer indexing on a NumPy array (inspired by Zarr).
+
+    This is more onerous, but calls `x.__getitem__` only once.
+    """
+    # Inspired by Zarr's indexing.py (https://github.com/zarr-developers/zarr)
+    # Get rid of ellipsis
+    keep = da.slicing.normalize_index(keep, x.shape)
+    new_axes = tuple(n for n, k in enumerate(keep) if k is np.newaxis)
+    drop_axes = tuple(n for n, k in enumerate(keep) if _is_integer(k))
+    # Get rid of newaxis
+    keep = tuple(k for k in keep if k is not np.newaxis)
+    keep = ix_(keep, x.shape)
+    result = x[keep]
+    for ax in new_axes:
+        result = np.expand_dims(result, ax)
+    result = result.squeeze(axis=drop_axes)
+    return result
+
+
+def numpy_oindex_lite(x, keep):
+    """Perform outer indexing on a NumPy array (compact version).
+
+    This is more compact, but calls `x.__getitem__` `x.ndim` times.
+
+    It also assumes that `keep` contains no ellipsis to be as pure as possible.
+    """
+    if not isinstance(keep, tuple):
+        keep = (keep,)
+    dim = 0
+    result = x
+    for k in keep:
+        cumulative_index = (slice(None),) * dim + (k,)
+        result = result[cumulative_index]
+        # Handle dropped dimensions
+        if not _is_integer(k):
+            dim += 1
+    return result
+
+
+class TestNumPyOIndex(object):
+    """Test the :func:`~numpy_oindex` function."""
+    def setup(self):
+        shape = (10, 10, 10, 10)
```

Comment:
Use a different size on each axis so that if a code path that uses the size on a dimension picks the wrong dimension it will get picked up.

Reply:
Done.